### PR TITLE
OpenStack Content Delivery V1 support (SDK V2)

### PR DESCRIPTION
**Pull request by sharwell**
# Work in Progress

This pull request adds support for the basic functionality of the OpenStack Content Delivery Service V1 to the V2 SDK.

### CODE REVIEW

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+            Dictionary<string, string> parameters = new Dictionary<string, string>();
+            return GetBaseUriAsync(cancellationToken)
+                .Then(PrepareRequestAsyncFunc(HttpMethod.Get, template, parameters, cancellationToken))
+                .Select(task => new GetHomeApiCall(CreateJsonApiCall<HomeDocument>(task.Result)));
+        }
+
+        /// <inheritdoc/>
+        public virtual Task<GetHealthApiCall> PrepareGetHealthAsync(CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs]: :exclamation: This needs to be implemented.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        }
+
+        /// <inheritdoc/>
+        public virtual Task<GetSubsystemHealthApiCall> PrepareGetSubsystemHealthAsync(SubsystemId subsystemId, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs]: :exclamation: This needs to be implemented.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+            return GetBaseUriAsync(cancellationToken)
+                .Then(PrepareRequestAsyncFunc(HttpMethod.Post, template, parameters, serviceData, cancellationToken))
+                .Select(task => new AddServiceApiCall(CreateCustomApiCall(task.Result, HttpCompletionOption.ResponseContentRead, deserializeResult)));
+        }
+
+        /// <inheritdoc/>
+        public virtual Task<UpdateServiceApiCall> PrepareUpdateServiceAsync(ServiceId serviceId, ServiceData serviceData, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs]: :exclamation: This needs to be implemented.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+            Dictionary<string, string> parameters = new Dictionary<string, string> { { "service_id", serviceId.Value } };
+            return GetBaseUriAsync(cancellationToken)
+                .Then(PrepareRequestAsyncFunc(HttpMethod.Delete, template, parameters, cancellationToken))
+                .Select(task => new RemoveServiceApiCall(CreateBasicApiCall(task.Result)));
+        }
+
+        /// <inheritdoc/>
+        public virtual Task<RemoveAssetApiCall> PrepareRemoveAssetAsync(ServiceId serviceId, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs]: :exclamation: This needs to be implemented.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+            return GetBaseUriAsync(cancellationToken)
+                .Then(PrepareRequestAsyncFunc(HttpMethod.Get, template, parameters, cancellationToken))
+                .Select(task => new ListFlavorsApiCall(CreateCustomApiCall(task.Result, HttpCompletionOption.ResponseContentRead, deserializeResult)));
+        }
+
+        /// <inheritdoc/>
+        public virtual Task<GetFlavorApiCall> PrepareGetFlavorAsync(FlavorId flavorId, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs]: :exclamation: This needs to be implemented.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+
+            return GetBaseUriAsync(cancellationToken)
+                .Then(PrepareRequestAsyncFunc(HttpMethod.Post, template, parameters, flavorData, cancellationToken))
+                .Select(task => new AddFlavorApiCall(CreateCustomApiCall(task.Result, HttpCompletionOption.ResponseContentRead, deserializeResult)));
+        }
+
+        /// <inheritdoc/>
+        public virtual Task<RemoveFlavorApiCall> PrepareRemoveFlavorAsync(FlavorId flavorId, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryClient.cs]: :exclamation: This needs to be implemented.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryServiceExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        {
+            if (service == null)
+                throw new ArgumentNullException("service");
+
+            return TaskBlocks.Using(
+                () => service.PrepareGetHomeAsync(cancellationToken),
+                task => task.Result.SendAsync(cancellationToken).Select(innerTask => innerTask.Result.Item2));
+        }
+
+        //public static Task<GetHealthApiCall> GetHealthAsync(this IContentDeliveryService service, CancellationToken cancellationToken)
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryServiceExtensions.cs]: :exclamation: These need to be implemented, or removed.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryServiceExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+    /// <summary>
+    /// This class provides extension methods that simplify the process of preparing and sending Content Delivery
+    /// Service HTTP API calls for the most common use cases.
+    /// </summary>
+    /// <threadsafety static="true" instance="false"/>
+    /// <preliminary/>
+    public static class ContentDeliveryServiceExtensions
+    {
+        public static Task<HomeDocument> GetHomeAsync(this IContentDeliveryService service, CancellationToken cancellationToken)
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ContentDeliveryServiceExtensions.cs]: :exclamation: This needs to be documented (along with several other cases in the same file).

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/FlavorData.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FlavorData"/> class
+        /// during JSON deserialization.
+        /// </summary>
+        [JsonConstructor]
+        protected FlavorData()
+        {
+        }
+
+        public ImmutableArray<FlavorProvider> Providers
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/FlavorData.cs]: :exclamation:  This needs to be documented.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/IContentDeliveryService.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    /// <summary>
+    /// This is the base interface for the OpenStack Content Delivery Service V1.
+    /// </summary>
+    /// <seealso href="">OpenStack Content Delivery API V1 Reference</seealso>
+    /// <preliminary/>
+    public interface IContentDeliveryService : IHttpService, IExtensibleService<IContentDeliveryService>
+    {
+        #region Service Information
+
+        Task<GetHomeApiCall> PrepareGetHomeAsync(CancellationToken cancellationToken);
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/IContentDeliveryService.cs]: :exclamation: This needs to be documented (along with several other cases in the same file).

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/Service.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
+﻿namespace OpenStack.Services.ContentDelivery.V1
+{
+    using System.Collections.Immutable;
+    using Newtonsoft.Json;
+    using OpenStack.Services.Identity;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class Service : ServiceData
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/Service.cs]: :exclamation: This needs to be documented (along with several other elements in the same file).

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceCache.cs [hunk 12]
```diff
@@ -0,0 +1,73 @@
+﻿namespace OpenStack.Services.ContentDelivery.V1
+{
+    using System;
+    using System.Collections.Immutable;
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceCache : ExtensibleJsonObject
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceCache.cs]: :exclamation: This needs to be documented (along with several other elements in the same file).

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceCacheRule.cs [hunk 13]
```diff
@@ -0,0 +1,51 @@
+﻿namespace OpenStack.Services.ContentDelivery.V1
+{
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    public class ServiceCacheRule : ExtensibleJsonObject
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceCacheRule.cs]: :exclamation: This needs a `[JsonObject]` attribute.
:exclamation: This needs to be documented (along with several other elements in the same file).

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceData.cs [hunk 14]
```diff
@@ -0,0 +1,113 @@
+﻿namespace OpenStack.Services.ContentDelivery.V1
+{
+    using System.Collections.Immutable;
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceData : ExtensibleJsonObject
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceData.cs]: :exclamation: This needs to be documented (along with several other elements in the same file).

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceDomain.cs [hunk 15]
```diff
@@ -0,0 +1,66 @@
+﻿namespace OpenStack.Services.ContentDelivery.V1
+{
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceDomain : ExtensibleJsonObject
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceDomain.cs]: :exclamation: This needs to be documented.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceError.cs [hunk 16]
```diff
@@ -0,0 +1,34 @@
+﻿namespace OpenStack.Services.ContentDelivery.V1
+{
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceError : ExtensibleJsonObject
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceError.cs]: :exclamation: This needs to be documented.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceError.cs [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ServiceError"/> class
+        /// during JSON deserialization.
+        /// </summary>
+        [JsonConstructor]
+        protected ServiceError()
+        {
+        }
+
+        public string Message
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceError.cs]: :exclamation: This needs to be documented.

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceOrigin.cs [hunk 18]
```diff
@@ -0,0 +1,83 @@
+﻿namespace OpenStack.Services.ContentDelivery.V1
+{
+    using System.Collections.Immutable;
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceOrigin : ExtensibleJsonObject
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceOrigin.cs]: :exclamation: This needs to be documented (along with several other elements in the same file).

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceOriginRule.cs [hunk 19]
```diff
@@ -0,0 +1,51 @@
+﻿namespace OpenStack.Services.ContentDelivery.V1
+{
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    public class ServiceOriginRule : ExtensibleJsonObject
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceOriginRule.cs]: :exclamation: This needs to be documented (along with several other elements in the same file).

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceRestriction.cs [hunk 20]
```diff
@@ -0,0 +1,53 @@
+﻿namespace OpenStack.Services.ContentDelivery.V1
+{
+    using System.Collections.Immutable;
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceRestriction : ExtensibleJsonObject
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceRestriction.cs]: :exclamation: This needs to be documented (along with several other elements in the same file).

File: src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceRestrictionRule.cs [hunk 21]
```diff
@@ -0,0 +1,52 @@
+﻿namespace OpenStack.Services.ContentDelivery.V1
+{
+    using Newtonsoft.Json;
+    using OpenStack.ObjectModel;
+
+    [JsonObject(MemberSerialization.OptIn)]
+    public class ServiceRestrictionRule : ExtensibleJsonObject
```
Comment (sharwell) [inline src/OpenStack.Net/OpenStack/Services/ContentDelivery/V1/ServiceRestrictionRule.cs]: :exclamation: This needs to be documented (along with several other elements in the same file).

File: src/OpenStackNetTests.Live/ContentDeliveryTests.Shared.cs [hunk 22]
```diff
@@ -0,0 +1,405 @@
+﻿namespace OpenStackNetTests.Live
```
Comment (sharwell) [inline src/OpenStackNetTests.Live/ContentDeliveryTests.Shared.cs]: :exclamation: This file is currently only included in the Live tests project. It should be linked to the Unit tests project so a service simulator can be implemented for efficient unit testing of the service functionality.

**Comment by DonSchenck** [comment]
I'm going to work on implementing the above methods.

**Comment by sharwell** [comment]
@DonSchenck Awesome. I look forward to seeing the update! :+1:

**Comment by DonSchenck** [comment]
Why is this tagged as Do Not Merge? Let's merge it and then I can pull it down and continue work.

**Comment by sharwell** [comment]
In the work so far on the V2 SDK, we've only merged pull requests once they reach a certain level of completeness. In particular, while the functionality could be a subset of the complete offerings of the service, we looked for the following:

1. Implementations of code in the service client class.
2. Reasonably complete documentation.

We also looked for a minimal level of live testing coverage (focused on the most important API features).

There are two ways for multiple users to collaborate on a feature in this environment. To date, we've been using the first method.

1. The user who starts the work creates a feature branch in their fork of this repository. In the case of this pull request, the branch was created as **sharwell/v2-content-delivery**. Other users working on the same feature would add a new remote for the fork of the repository where the feature branch is. For example, the following adds a remote, checks out the feature branch, and sets your copy of the branch to upload to *your* fork of the project from which you can send a pull request:

    ```bash
    # Add a remote 'sharwell', where the feature branch is located (only need to do this once
    # per remote)
    git remote add sharwell https://github.com/sharwell/openstack.net.git
    # Fetch from the 'sharwell' remote, so your local copy has up-to-date information about its
    # working branches
    git fetch sharwell
    # Check out a feature branch from the 'sharwell' remote
    git checkout -b v2-content-delivery sharwell/v2-content-delivery
    # Set your local copy of the feature branch to track 'origin' remote instead of 'sharwell'.
    # This will push to your fork, and then you can submit a pull request to the original feature
    # branch in the 'sharwell' remote. When that pull request is merged into the feature branch,
    # *this* pull request (#475) will automatically be updated to include that work.
    git branch -u origin/v2-content-delivery v2-content-delivery
    ```

    To submit work to this feature branch prior to *this* pull request (#475) getting merged, a user can submit a pull request to the feature branch (see [Changing the branch range and destination repository](https://help.github.com/articles/using-pull-requests/#changing-the-branch-range-and-destination-repository) for more information on sending a pull request to a specific feature branch in a specific fork).

2. The alternative is to create feature branches in the parent repository. The workflow for this is essentially the same as for the first step, except users do not need to use `git remote add` and the intermediate pull requests only need to specify a branch (because the default repository is the correct target). The primary downside to this approach is it tends to lead to many branches in the main repository, which can be challenging for new users to sort through.

**Comment by DonSchenck** [comment]
Got this:
```
C:\Users\User\Documents\GitHub\openstack.net [v2-content-delivery +1 ~1 -1 !]> git branch -u origin/v2-content-delivery v2-content-delivery
error: the requested upstream branch 'origin/v2-content-delivery' does not exist
hint:
hint: If you are planning on basing your work on an upstream
hint: branch that already exists at the remote, you may need to
hint: run "git fetch" to retrieve it.
hint:
hint: If you are planning to push out a new local branch that
hint: will track its remote counterpart, you may want to use
hint: "git push -u" to set the upstream config as you push.
```

**Comment by sharwell** [comment]
@DonSchenck Thanks, I wasn't sure if that would happen. I changed that step to be two different commands in the original example. The `push` command I added at the end creates the remote branch mentioned in that error at the same time it assigns the tracking branch.

**Comment by sharwell** [comment]
Superseded by #500.

**Pull request closed without merging**